### DISCUSSION TITLE
Rekonfiguration abgelehnt - Es ist Ihnen nicht erlaubt, SEB zu rekonfigurieren

### DISCUSSION THREAD

**Discussion by FH-Campuswien-HR** [SafeExamBrowser/seb-win-refactoring#433] [Q&A]
Wir haben an der FH Campus Wien verschiedene, örtlich getrennte, Standorte. Prüfungen finden mittels SEB (Version 3.2.2, Windows) und Moodle (Version 3.11.5+) in den EDV Lehrsälen an den jeweiligen Standorten statt. Zur Konfiguration des SEB verwenden wir in Moodle die Option mittels Konfigurations-Template die Konfiguration vorzunehmen. 

Bei Prüfungen die aus einem einzelnen Test bestehen funktioniert das absolut problemlos. Jedoch bei Prüfungen die aus mehreren Tests bestehen, die hintereinander absolviert werden müssen, kommt es zu Problemen:
Der erste Test kann problemlos absolviert werden und auch das zurückkehren zur Kursübersicht ist kein Problem. Wenn man dann aber den nächsten Test starten möchte kommt eine Fehlermeldung („Rekonfiguration abgelehnt - Es ist Ihnen nicht erlaubt, SEB zu rekonfigurieren.“). Die einzige Möglichkeit ist, den SEB zu schließen und neu zu starten um den nächsten Test aufrufen zu können.

Das Problem tritt nicht auf, wenn für die Konfiguration des SEB statt dem Template über Moodle eine lokal gespeicherte SEB-Konfigurationsdatei verwendet wird. Dann können alle Tests hintereinander problemlos absolviert werden.

Wir bitte um Unterstützung bei der Lösung des Problems bzw. um Auskunft ob es überhaupt ein Problem ist oder ob diese Einschränkung sein muss, aus Gründen die uns einfach nicht bekannt sind. 

Vielen Dank!

Sreenshot Fehlermeldung:
![Screenshot-SEB Rekonfiguration abgelehnt](https://user-images.githubusercontent.com/108469569/176699949-d7e80bab-ab46-4293-8588-f8cfb86d1476.jpg)

**Comment by heinzkrebs** [comment]
Ich nehme an, dass mit jedem Starten eines Tests auch die SEB-konfigurationsdatei neu geladen wird. Allerdings gibt es ein Setting, welches dies verhindert/erlaubt.

Moodle (zumindest in der TUWien-Version) hat keine Option dieses Setting zu konfigurieren, also ist es vielleicht defaultmässig nicht erlaubt. In der lokal gespeicherten Konfigurationsdatei ist das eventuell erlaubt (im Tab Exam links unten gibt es das Setting "Reconfigure Secure/Exam Session")

mfg
 H.

**Reply by danschlet** [reply]
Ja, die Einstellung "Allow reconfiguring secure/exam session" aus dem [Exam Tab](https://safeexambrowser.org/windows/win_usermanual_en.html#ExamPane) müsste benutzt werden, und bei "Reconfiguring URL" sollte der Pfad der Moodle-Installation angegeben mit /* am Ende (oder einfacher direkt nur das Wildcard-Sternzeichen *). Damit müsste man eine entsprechende SEB-Konfigurationsdatei im SEB Configuration Tool erstellen und diese dann mit der Option Require Safe Exam Browser "Yes - Upload my own config" nutzen. Allerdings müsste auch die Konfiguration für das erste Quiz bereits entsprechend angepasst sein, die Einstellung "Clear browser session when ending an exam or terminating SEB" muss deaktiviert sein, bei der zweiten oder jeder weiteren Konfiguration muss "Clear browser session when starting an exam or starting SEB" deaktiviert werden, damit die Studierenden nicht ausgeloggt werden beim Starten eines weiteren Quizzes.

Eine andere Möglichkeit wäre, SEB manuell (beispielsweise über einen sebs:// Link) mit einer Konfigurationsdatei zu starten, welche als Start URL die Moodle-Kursübesichtsseite enthält. Bei den Quizzes, die man dann entsprechend verwenden will, kann man dann die Option "Yes - use Client Config" nutzen, um diese aber gegen unbefugten Zugriff abzusichern, müsste man die Browser Exam Key (BEK) Zeichenkette aus dem Configuration Tool / Exam Tab für die erwähnte Konfigurationsdatei in das entsprechende Feld der Quiz-Einstellungen kopieren. Der BEK ist jeweils versionsabhängig, zum Aufsetzen und zur Prüfungsdurchführung muss also jeweils genau dieselbe SEB-Version benutzt werden (oder wenn man mehrere Versionen erlauben will, muss man den BEK für jede Version in die Quiz-Einstellungen kopieren).

**Reply by danschlet** [reply]
Übrigens erlaubt SEB das Umkonfigurieren einer sicheren Session absichtlich nicht. Man muss aufpassen, da es bei Verwendung einer Freitextfrage, in welcher der HTML-Editor erlaubt ist, theoretisch möglich wäre, einen SEB-Konfigurationslink zu erstellen, mit welchem dann SEB während der Prüfung auf unsichere Einstellungen umkonfiguriert würde. Das selbe gilt, wenn Drittapplikationen oder externe Webseiten in Prüfungen benutzt werden können, die es ermöglichen, SEB-Konfigurationsdateien oder -Links zu erstellen. Allerdings würde der Zugriff auf die Prüfung beim Umkonfigurieren eines laufenden Moodle-Quizzes, für welches eingestellt ist, dass die Verwendung von SEB erforderlich ist, verweigert, weil sich eben die Konfiguration geändert hat. Es gibt trotzdem Szenarien, bei welchen man damit mogeln könnte, wenn die Prüfungsaufsicht nicht genug aufmerksam wäre.

**Reply by AntonT76** [reply]
Hallo @danschlet ,

wir haben das gleiche Problem mit den zwei Quizzes wie @FH-Campuswien-HR  und haben es nun gemäß deinen Empfehlungen im ersten Absatz gelöst. Ich habe aber ein paar Fragen zur Einstellung:

a.) Clear browser session when starting an exam or starting SEB
b.) Clear browser session when ending an exam or terminating SEB (prevents deletion of browser cache if deactivated)
c.) Allow reconfiguring secure/exam session

Wenn ich a. und b. deaktiviere und dafür c. aktiviere und das Setting bei beiden Quizzes verwende, dann wird beim zweiten Quiz zwar der SEB neu gestartet, aber ein erneuter Login ist nicht notwendig. Im Grunde müsste daher ein einziges SEB-File mit diesen Einstellungen für alle aufeinanderfolgenden Quizzes ausreichend sein. Wenn ich deine Beschreibung im Absatz 1 richtig interpretiere, müssten sich ja die Studierenden beim Starten des 2. Quizzes erneut einloggen müssen - das ist bei mir nicht der Fall. Oder hab ich das falsch verstanden?

Zu c. noch eine Anmerkung: ich hab ja in einem anderen Posting (https://github.com/SafeExamBrowser/seb-win-refactoring/discussions/435) darüber berichtet, dass ich mich noch ein drittes Mal einloggen muss. Wenn ich c. aktiviere, a. und b. deaktiviere, dann muss ich mich am Beginn "nur" 2x einloggen. Vielleicht hilft das bei der Fehlersuche zum 3. Login weiter.

BG Anton

**Reply by danschlet** [reply]
> Wenn ich deine Beschreibung im Absatz 1 richtig interpretiere, müssten sich ja die Studierenden beim Starten des 2. Quizzes erneut einloggen müssen - das ist bei mir nicht der Fall. 

Wenn alle Konfigurationen sowohl a) als auch b) deaktiviert haben (was aber nicht die Standardeinstellung ist), dann müsste man sich auch beim Starten des 2. Quizzes nicht erneut einloggen. Nur wenn man als erstes ein Quiz mit Standardeinstellung (also z.B. eines mit der Moodle-Option YES - manual...) für diese zwei Settings verwendet, dann wird man sich beim zweiten erneut einloggen müssen.
Wenn sich ein User einloggt, dann wird von Moodle ein Session-Cookie im Browser gesetzt. Normalerweise würde SEB diese Session-Cookies beim Start als auch beim Beenden einer Session löschen (äquivalent wenn man einen normalen Browser beendet und neu startet), mit diesen zwei Einstellungen kann man die Session-Cookies zu Beginn oder zum Ende einer Session behalten. Wichtig zu wissen ist allerdings, dass die Session-Cookies bei der Windows-Version aus technischen Gründen eben auch beim kompletten Beenden von SEB erhalten bleiben, wenn Option b) angewählt war. Man könnte sich also ein Sicherheitsproblem schaffen, wenn man diese Optionen grundsätzlich deaktiviert (aufpassen muss man wohl vor allem bei Computern, die von mehreren Studierenden genutzt werden, bei BYOD sollte das kein Problem sein). Sorry, das ist leider ein wenig komplex, lässt sich aber für den Use Case von aufeinanderfolgenden Prüfungen kaum vermeiden.

**Reply by AntonT76** [reply]
Hallo @danschlet 

sorry, dass ich da noch nachhake - ich hab noch eine Anmerkung zu den beiden Optionen unter Session Handling:
a.) Clear browser session when starting an exam or starting SEB
b.) Clear browser session when ending an exam or terminating SEB (prevents deletion of browser cache if deactivated)

Man könnte ja eine zusätzliche Option zur Auswahl im SEB Config Tool einbauen, die diese beiden Funktionen mit dem aktiven Schließen des SEB durch die User auslöst (Button: Terminate Session). Dann würden die Browser Sessions innerhalb mehrerer aufeinanderfolgenden Klausuren des Users zwar bestehen bleiben, wenn aber ein User den SEB schließt, erfolgt ein Löschen der Sessions. Zumindest bei uns es es das Szenario, dass User nach der Prüfung den SEB schließen (sie sind ja mit ihrem persönlichen Userkonto am Rechner angemeldet und da ist es auch ein persönliches Interesse der User, sich auch vom Rechner auszuloggen). Das würde zumindest das Sicherheitsproblem zusätzlich minimieren.

BG Anton

**Reply by danschlet** [reply]
Das Problem bei der Windows-Version ist, dass deren Architektur das generelle Löschen der Cookies beim Beenden von SEB (unabhängig von den erwähnten Settings) schwierig zu implementieren macht. Ich sehe an sich keinen Use Case, wo die Session Cookies beim Beenden der Applikation erhalten bleiben müssten. Ich werde das mit dem Entwickler @dbuechel nochmals anschauen, vielleicht gibt es ja doch eine Möglichkeit, das so zu implementieren wie bei der Mac- oder iOS-Version.

**Reply by AntonT76** [reply]
ok danke.

Ich ergänze mal einen Use Case - auch wenn der sehr speziell auf unsere Bedürfnisse zugeschnitten ist:
- wir möchten aufgrund der umfangreichen Einstellungsmöglichkeiten im SEB den Lehrenden die Konfiguration des SEB abnehmen und vorgefertigte, standardisierte SEB-Files zur Verfügung stellen, die die Lehrenden dann nur noch in den Test hochladen müssen
- bei aufeinanderfolgenden Tests ist die Fehlerquelle bei den Einstellungen im Session Handling relativ groß (z. B. Verwechslung der SEB-Files; was ist, wenn jemand 3 aufeinanderfolgende Tests hat, ...).
- Vermeidung des zusätzlichen Logins: wir arbeiten mit einer Mehrfaktor-Authentifizierung - zu Beginn der Prüfung dürfen die User noch ihr Mobiltelefon als 2. Faktor verwenden, danach müssen sie es weglegen. Bei aufeinanderfolgenden Tests möchten wir vermeiden, dass es zu einem nochmaligen Login und erneuter Nutzung des Mobiltelefons während der Prüfung kommt

BG Anton

**Comment by FH-Campuswien-HR** [comment]
Alles klar, vielen Dank für die Antworten, dass hat mir schon sehr weitergeholfen!
